### ISSUE TITLE
replace StopTraversal with () or []

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11016]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| andrea reported|
|-|-|
|Trac ID|trac#1471|
|Type|enhancement|
|Created|2006-02-07 00:55:48Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1471 1471
type__enhancement enhancement
reporter__andrea andrea
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__wontfix wontfix
component__web2 web2
keywords__ 
time__1139273748000000 1139273748000000
changetime__1143139646000000 1143139646000000
version__ 
owner__ 
cc__jknight cc__andrea cc__dreid
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
I don't like StopTraversal, the self documenting API for the segment is to be   
popped from the head until it becomes an empty list or an empty tuple.  
StopTraversal is an unnecessary complication to the traversal API.  
   
http://www.cpushare.com/hg/Twisted?cs=d11429a2621a  
 
Thanks.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

```
#!html
<pre>
resolved at r15857
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
I don't think this is right. "return self, StopTraversing" was used to keep the 
remaining path segments in the req.postpath for further processing when renderHTTP 
was invoked. "return self, ()" caused renderHTTP to be called with self.postpath 
== [].
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

```
#!html
<pre>
I thought about that, and decided it'd be better to let the Resource handle
remembering of consumed path segments.  Apparently I dreamed mentioning this to
you (i guess that's what the tracker is for.)  Anyway I'm of the opinion that
the resource that returns () should be in charge of handling arbitrarily
consumed path segments and getting them to where they need to be.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
I don't know about that: it was always a pain in the ass in Nevow that you 
couldn't easily let your renderer handle the remaining path segments.

Additionally, the change is clearly wrong: the docstring for locateChild is now 
a lie, and the places that used StopTraversal were depending upon its difference 
in behavior vs () are now broken. I'll take the blame for not having written 
enough tests, but still.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
Are you really sure you want to allow render to be invoked before postpath is 
empty? That doesn't look a strightforward API to me. The traversal should stop 
only when all segments have been processed. _After_ the traversal is complete, 
the processing can start. 
 
What's the point anyway to have the info in postpath when you can find all 
information in req.prepath inside the rendering callback? If you can tell me a 
single reason why you can't find the very same information you have in 
postpath, in the prepath, then I could reconsider. 
 
Now if prepath is losing data, then fix that, since that is the real problem. 
 
I generally dislike unnecessary special cases that breaks invariants, and this 
is one of them as far as I can tell.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

```
#!html
<pre>
I don't think req.postpath should be the api for accessing "paths I have
consumed."  Because req.postpath is really 'paths that have not been processed'.
 If you stop traversal you're processing or atleast performing an action
(consuming) the rest of the paths, so _I'd_ expect req.postpath to be empty.

Perhaps LeafResource could grow a 'consumedPaths' attribute that is set with the
remaining segments.  I don't think it's unreasonable to require a resource to
keep track of it's own consumed paths, it's more symmetrical with the way you
have to deal with consuming arbitrary path segments (though not necessarily all
of the path segments.)
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
David I agree it's more natural and cleaner API to stop the traversal only     
after postpath has been fully processed.    
     
I'm not sure to understand the detail of consumedPaths, but I did a test. I've   
a couple of pages in CPUShare parsing the 'segments' parameter of locateChild   
with sql queries. I process them in the locateChild and then I store the   
related information in a req.private class. Let's assume I wanted to get rid   
of the req.private class and I want to lookup again the db in the render   
callback (I understand that's the objective of the StopIteration), the   
information is still there in req.prepath, I added a "print req.prepath" and   
all segments were there.   
   
So if you want to process them in render() instead of locateChild it's still   
possible.   
   
Problem is for me it's more efficient to query the db only once in locateChild   
and to pass up the result to the render() method of the page, through a   
req.private field. I am forced to query the db once to know which is the   
render method to invoke, and so I grab all related data at the same time.   
StopIteration would make no difference, since you would still be forced to   
query the db once in locateChild. So whatever StopIteration could do, can also   
be done with req.prepath and it is an overall suboptimal model.   
   
I'd rather prefer us to work to have a more standardized way of passing 
information from thisclass.locateChild to the otherclass.render() method of 
the resulting page. My way of using a req.private class isn't the cleanest (of 
course I could use zope interfaces to avoid the namespace collisions, but I 
generally hate the interface inefficient bloat, I rather prefer a full 
breakage with namespace collision, full breakages are generally ok because 
they're trivial to fix).  
 
For the session I do it very similarly, with req.session = session_class().   
  
here an example:  
   
 			class private(object):  
				def __init__(self, email):  
					self.email = email  
			req.private = private(result[0][0])  
			return verify_email2_class(), remaining_segments
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

```
#!html
<pre>
def locateChild(self, req, seg):
    self.consumedPaths = seg
    return self, ()

def render(self, req):
    do things with self.consumedPaths if necessary.

That was the point of my consumedPaths suggestion.  getting the consumed paths
just the consumed paths, not paths before our current location in the tree to
the _current_ resource.

Standardized way of passing information from thisresrc to thatresrc when
thisresrc.locateChild returns an instance of thatresrc would be this.

def locateChild(self, req, seg):
    r = result of something
    return thatresrc(r), seg[1:]

foom perhaps we should move this discussion to the mailing list to get more user
input?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
So I understood your consumedPaths right...  Well, it's a by-design racy, so I   
think it's unacceptable as generic API. The race happens if the same instance   
can be used by multiple req and you can reschedule during locatechild or   
during render. There has never been the requirement that a new instance of the   
Resource classes has to be created for every new request. Otherwise obviously   
you should set the request in Resource.req (self.req = req inside __init__)   
and not passed as parameter to every method (locateChild/render).   
   
About the API, to pass info from locateChild to render() that has a similar   
issue, but a least if the information is forced to be set in __init__, subtle   
races are less likely since you're be basically forced to create a new   
instance for every different req by design.   
   
But the consumedPaths didn't force it, this is why it's much less acceptable 
than passing the parameters through __init__. 
 
My suggestion was to have an API inside the request, exactly to follow the non 
racy model, where new instances of resources don't necessarily have to be 
created for every new incoming request.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="andrea's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|andrea commented|
|-|-|

```
#!html
<pre>
I had to back it out temporarily because I found the hard way it broke scgi, I  
didn't notice this specific LeafResource subclass depended on the postpath  
behaviour.  
  
 http://www.cpushare.com/hg/Twisted?cs=591aaa82f2b9 
 
Even though I had to back it out, my sentiment didn't change much about 
StopTraversal, and if we fix scgi to work with current SVN, we'll also see how 
much uglier the code becomes without special cases in the path traversal. 
 
If the code becomes too ugly at least we'll be sure that the StopTraversal 
special case was there for good ;)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set status to `closed`|
|-|-|

After some long discussions with jknight on this subject I decided that StopTraversal was right and my removal of it and support of this bug was based on my own missunderstanding.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
